src/rp2_common/hardware_gpio/gpio.c

```diff
@@ -178,21 +183,46 @@ void gpio_set_irq_enabled(uint gpio, uint32_t events, bool enabled) {
 
 void gpio_set_irq_enabled_with_callback(uint gpio, uint32_t events, bool enabled, gpio_irq_callback_t callback) {
     gpio_set_irq_enabled(gpio, events, enabled);
+    gpio_set_irq_callback(callback);
+    if (enabled) irq_set_enabled(IO_IRQ_BANK0, true);
+}
+
+void gpio_set_irq_callback(gpio_irq_callback_t callback) {
+    uint core = get_core_num();
+    if (callbacks[core]) {
+        if (!callback) {
+            irq_remove_handler(IO_IRQ_BANK0, gpio_default_irq_handler);
+        }
+        callbacks[core] = callback;
+    } else if (callback) {
+        callbacks[core] = callback;
+        irq_add_shared_handler(IO_IRQ_BANK0, gpio_default_irq_handler, GPIO_IRQ_CALLBACK_ORDER_PRIORITY);
+    }
+}
+
+void gpio_add_raw_irq_handler_with_order_priority_masked(uint gpio_mask, irq_handler_t handler, uint8_t order_priority) {
```

Comment:
Should we assert that there is not already a raw IRQ for this GPIO?

Reasoning is that we unconditionally clear the raw bit in the `remove` function, so going `add` -> `add` -> `remove` on the same GPIO leaves it marked as non-raw even though it does have a raw handler installed.